api/internal/oci/repospec.go

```diff
@@ -0,0 +1,246 @@
+
+	// Parse the host (e.g. scheme, domain) segment.
+	repoSpec.Host, n, err = extractHost(n)
+	if err != nil {
+		return nil, err
+	}
+
+	repoSpec.KustRootPath, n, err = extractRoot(n)
+	if err != nil {
+		return nil, err
+	}
+
+	repoSpec.Digest, n, err = extractDigest(n)
+	if err != nil {
+		return nil, err
+	}
+
+	repoSpec.Tag, repoSpec.RepoPath, err = extractTag(n)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(repoSpec.RepoPath) == 0 {
+		return nil, errors.Errorf("failed to parse repo path segment")
+	}
+
+	return repoSpec, nil
+}
+
+func extractRoot(n string) (string, string, error) {
+	if rootIndex := strings.LastIndex(n, rootSeparator); rootIndex >= 0 {
+		root := n[rootIndex+len(rootSeparator):]
+
+		if len(root) == 0 {
+			return "", "", errors.Errorf("failed to parse root path segment")
+		}
+		if kustRootPathExitsRepo(root) {
+			return "", "", errors.Errorf("root path exits repo")
+		}
+
+		return root, n[:rootIndex], nil
+	}
+
+	return "", n, nil
+}
+
+func kustRootPathExitsRepo(kustRootPath string) bool {
+	cleanedPath := filepath.Clean(strings.TrimPrefix(kustRootPath, string(filepath.Separator)))
+	pathElements := strings.Split(cleanedPath, string(filepath.Separator))
+	return len(pathElements) > 0 &&
+		pathElements[0] == filesys.ParentDir
+}
+
+// Arbitrary, but non-infinite, timeout for running commands.
+const defaultTimeout = 27 * time.Second
+
+func extractDigest(n string) (string, string, error) {
+	if digestIndex := strings.LastIndex(n, digestSeparator); digestIndex >= 0 {
+		digest := n[digestIndex+len(digestSeparator):]
+		// Digest is at least 8 characters, but we don't validate the entire schema here
+		if len(digest) < 8 {
+			return "", "", errors.Errorf("failed to parse digest")
+		}
+
+		return digest, n[:digestIndex], nil
+	}
+
+	// No digest
+	return "", n, nil
+}
+
+func extractTag(n string) (string, string, error) {
+	if tagIndex := strings.LastIndex(n, tagSeparator); tagIndex >= 0 {
+		tag := n[tagIndex+len(tagSeparator):]
+
+		if len(tag) == 0 {
+			return "", "", errors.Errorf("failed to parse tag segment")
+		}
+
+		return tag, n[:tagIndex], nil
+	}
+
+	return "", n, nil
+}
+
+func extractHost(n string) (string, string, error) {
+	scheme, n, err := extractScheme(n)
+
+	if err != nil {
+		return "", "", err
+	}
+
+	// Now that we have extracted a valid scheme, we can parse host itself.
+
+	// The oci layout specifies a path to a local oci layout directory or archive.
+	// Everything after the scheme is actually part of that path.
+	if scheme == ociLayoutScheme {
+		return ociLayoutScheme, n, nil
+	}
+
+	var host, rest = n, ""
+	if sepIndex := findPathSeparator(n); sepIndex >= 0 {
+		host, rest = n[:sepIndex+1], n[sepIndex+1:]
+	}
+
+	// Host is required, so do not concat the scheme and username if we didn't find one.
+	if host == "" {
+		return "", "", errors.Errorf("failed to parse host segment")
+	}
+	return host, rest, nil
+}
+
+const ociScheme = "oci://"
+const ociLayoutScheme = "oci-layout://"
```

Comment:
Are people really using `oci-layout://` for a scheme here? Does flux support that?

Reply:
I don't think I've seen an official "scheme" for this here.  ORAS itself has CLI flags for remote vs layout directories/archives (as do some other tools), but that's not really an option in this case.

Flux doesn't appear to support OCI Layout artifacts currently.